src/lib/optimizer/chunk_statistics.hpp

```diff
@@ -0,0 +1,170 @@
+
+  /**
+   * checks whether the filter is able to determine that the given value 
+   * and predicate condition will not yield any positive results with the values
+   * represented by the filter data.
+   * 
+   * In other words: A scan operation with value and predicate_type on the column/chunk
+   * that this filter was created on would yield zero result rows.
+  */
+  virtual bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const = 0;
+};
+
+class ChunkColumnStatistics {
+ public:
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics(DataType data_type,
+                                                                 std::shared_ptr<BaseColumn> column);
+
+  void add_filter(std::shared_ptr<BaseFilter> filter) { _filters.emplace_back(filter); }
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const {
+    for (const auto& filter : _filters) {
+      if (filter->can_prune(value, predicate_type)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+ protected:
+  std::vector<std::shared_ptr<BaseFilter>> _filters;
+};
+
+template <typename T>
+class MinMaxFilter : public BaseFilter {
+ public:
+  explicit MinMaxFilter(T min, T max) : _min(min), _max(max) {}
+  ~MinMaxFilter() override = default;
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const override {
+    const auto t_value = type_cast<T>(value);
+    // Operators work as follows: value_from_table <operator> t_value
+    // e.g. OpGreaterThan: value_from_table > t_value
+    // thus we can exclude chunk if t_value >= _max since then no value from the table can be greater than t_value
+    switch (predicate_type) {
+      case PredicateCondition::GreaterThan:
+        return t_value >= _max;
+      case PredicateCondition::GreaterThanEquals:
+        return t_value > _max;
+      case PredicateCondition::LessThan:
+        return t_value <= _min;
+      case PredicateCondition::LessThanEquals:
+        return t_value < _min;
+      case PredicateCondition::Equals:
+        return t_value < _min || t_value > _max;
+      default:
+        return false;
+    }
+  }
+
+ protected:
+  const T _min;
+  const T _max;
+};
+
+template <typename T>
+class RangeFilter : public BaseFilter {
+ public:
+  static_assert(std::is_arithmetic_v<T>, "RangeFilter should not be instantiated for strings.");
+
+  explicit RangeFilter(std::vector<std::pair<T, T>> ranges) : _ranges(std::move(ranges)) {}
+  ~RangeFilter() override = default;
+
+  static std::unique_ptr<RangeFilter<T>> build_filter(const pmr_vector<T>& dictionary);
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const override {
+    const auto t_value = type_cast<T>(value);
+    switch (predicate_type) {
+      case PredicateCondition::Equals: {
+        bool prunable = false;
+        for (const auto & [ min, max ] : _ranges) {
```

Comment:
Jenkins throws an error on this line:
``` In file included from ../../../src/lib/optimizer/chunk_statistics.cpp:1:

../../../src/lib/optimizer/chunk_statistics.hpp:107:27: error: unused variable '' [-Werror,-Wunused-variable]

        for (const auto & [ min, max ] : _ranges) {
```
although all values are in use.
I could not reproduce this on my machine, using Arch Linux and Clang 5.0.1
Anyone know whats wrong?

Reply:
I cannot reproduce it on my mac (Clang 5) either :-/

Reply:
This might be an issue with the compiler. At least that’s my guess when I see that there actually is no name of the supposedly unused variables.

You can get the exact CI setup if you run the docker container described in the repo. It uses the newest clang version from the apt repo, but it might not be the latest one.

Reply:
Of course, if we could find an actual, current compiler bug, that would be super cool ;)

Reply:
I can reproduce this on my mac.

Reply:
I changed this so the structured binding is located within the for loop. I want this to be working asap, since the pr has been open for over 1 month now.